In [1]:
from __future__ import print_function
import numpy as np
from keras.datasets import mnist
from keras.models import Sequential
from keras.layers.core import Dense, Activation
from keras.optimizers import SGD
from keras.utils import np_utils
np.random.seed(1671) # for reproducibility
# network and training
NB_EPOCH = 20
BATCH_SIZE = 128
VERBOSE = 1
NB_CLASSES = 10 # number of outputs = number of digits
OPTIMIZER = SGD() # optimizer, explained later in this chapter
N_HIDDEN = 128
VALIDATION_SPLIT=0.2 # how much TRAIN is reserved for VALIDATION
# data: shuffled and split between train and test sets
(X_train, y_train), (X_test, y_test) = mnist.load_data()
#X_train is 60000 rows of 28x28 values --> reshaped in 60000 x 784
RESHAPED = 784
#
X_train = X_train.reshape(60000, RESHAPED)
X_test = X_test.reshape(10000, RESHAPED)
X_train = X_train.astype('float32')
X_test = X_test.astype('float32')
# normalize
X_train /= 255
X_test /= 255
print(X_train.shape[0], 'train samples')
print(X_test.shape[0], 'test samples')
# convert class vectors to binary class matrices
Y_train = np_utils.to_categorical(y_train, NB_CLASSES)
Y_test = np_utils.to_categorical(y_test, NB_CLASSES)
# M_HIDDEN hidden layers
# 10 outputs
# final stage is softmax
model = Sequential()
model.add(Dense(N_HIDDEN, input_shape=(RESHAPED,)))
model.add(Activation('relu'))
model.add(Dense(N_HIDDEN))
model.add(Activation('relu'))
model.add(Dense(NB_CLASSES))
model.add(Activation('softmax'))
model.summary()
model.compile(loss='categorical_crossentropy',
optimizer=OPTIMIZER,
metrics=['accuracy'])
history = model.fit(X_train, Y_train,
batch_size=BATCH_SIZE, epochs=NB_EPOCH,
verbose=VERBOSE, validation_split=VALIDATION_SPLIT)
score = model.evaluate(X_test, Y_test, verbose=VERBOSE)
print("Test score:", score[0])
print('Test accuracy:', score[1])


Using TensorFlow backend.


11493376/11490434 [==============================] - 0s 0us/step
60000 train samples
10000 test samples
Model: "sequential_1"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_1 (Dense)              (None, 128)               100480    
_________________________________________________________________
activation_1 (Activation)    (None, 128)               0         
_________________________________________________________________
dense_2 (Dense)              (None, 128)               16512     
_________________________________________________________________
activation_2 (Activation)    (None, 128)               0         
_________________________________________________________________
dense_3 (Dense)              (None, 10)                1290      
_________________________________________________________________
activation_3 (Activation)    (None, 10)                0         
Total params: 11

In [2]:
from __future__ import print_function
import numpy as np
from keras.datasets import mnist
from keras.models import Sequential
from keras.layers.core import Dense, Activation
from keras.optimizers import SGD
from keras.utils import np_utils
np.random.seed(1671) # for reproducibility
# network and training
NB_EPOCH = 50
BATCH_SIZE = 128
VERBOSE = 1
NB_CLASSES = 10 # number of outputs = number of digits
OPTIMIZER = SGD() # optimizer, explained later in this chapter
N_HIDDEN = 128
VALIDATION_SPLIT=0.2 # how much TRAIN is reserved for VALIDATION
# data: shuffled and split between train and test sets
(X_train, y_train), (X_test, y_test) = mnist.load_data()
#X_train is 60000 rows of 28x28 values --> reshaped in 60000 x 784
RESHAPED = 784
#
X_train = X_train.reshape(60000, RESHAPED)
X_test = X_test.reshape(10000, RESHAPED)
X_train = X_train.astype('float32')
X_test = X_test.astype('float32')
# normalize
X_train /= 255
X_test /= 255
print(X_train.shape[0], 'train samples')
print(X_test.shape[0], 'test samples')
# convert class vectors to binary class matrices
Y_train = np_utils.to_categorical(y_train, NB_CLASSES)
Y_test = np_utils.to_categorical(y_test, NB_CLASSES)
# M_HIDDEN hidden layers
# 10 outputs
# final stage is softmax
model = Sequential()
model.add(Dense(N_HIDDEN, input_shape=(RESHAPED,)))
model.add(Activation('relu'))
model.add(Dense(N_HIDDEN))
model.add(Activation('relu'))
model.add(Dense(NB_CLASSES))
model.add(Activation('softmax'))
model.summary()
model.compile(loss='categorical_crossentropy',
optimizer=OPTIMIZER,
metrics=['accuracy'])
history = model.fit(X_train, Y_train,
batch_size=BATCH_SIZE, epochs=NB_EPOCH,
verbose=VERBOSE, validation_split=VALIDATION_SPLIT)
score = model.evaluate(X_test, Y_test, verbose=VERBOSE)
print("Test score:", score[0])
print('Test accuracy:', score[1])

60000 train samples
10000 test samples
Model: "sequential_2"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_4 (Dense)              (None, 128)               100480    
_________________________________________________________________
activation_4 (Activation)    (None, 128)               0         
_________________________________________________________________
dense_5 (Dense)              (None, 128)               16512     
_________________________________________________________________
activation_5 (Activation)    (None, 128)               0         
_________________________________________________________________
dense_6 (Dense)              (None, 10)                1290      
_________________________________________________________________
activation_6 (Activation)    (None, 10)                0         
Total params: 118,282
Trainable params: 118,282
Non-trainable params: 0
_________

Epoch 48/50
48000/48000 [==============================] - 5s 96us/step - loss: 0.0994 - accuracy: 0.9718 - val_loss: 0.1235 - val_accuracy: 0.9657
Epoch 49/50
48000/48000 [==============================] - 6s 119us/step - loss: 0.0976 - accuracy: 0.9723 - val_loss: 0.1203 - val_accuracy: 0.9666
Epoch 50/50
10000/10000 [==============================] - 0s 46us/step
Test score: 0.11470479362681509
Test accuracy: 0.9656999707221985


In [3]:
from __future__ import print_function
import numpy as np
from keras.datasets import mnist
from keras.models import Sequential
from keras.layers.core import Dense, Activation
from keras.optimizers import SGD
from keras.utils import np_utils
np.random.seed(1671) # for reproducibility
# network and training
NB_EPOCH = 150
BATCH_SIZE = 128
VERBOSE = 1
NB_CLASSES = 10 # number of outputs = number of digits
OPTIMIZER = SGD() # optimizer, explained later in this chapter
N_HIDDEN = 128
VALIDATION_SPLIT=0.2 # how much TRAIN is reserved for VALIDATION
# data: shuffled and split between train and test sets
(X_train, y_train), (X_test, y_test) = mnist.load_data()
#X_train is 60000 rows of 28x28 values --> reshaped in 60000 x 784
RESHAPED = 784
#
X_train = X_train.reshape(60000, RESHAPED)
X_test = X_test.reshape(10000, RESHAPED)
X_train = X_train.astype('float32')
X_test = X_test.astype('float32')
# normalize
X_train /= 255
X_test /= 255
print(X_train.shape[0], 'train samples')
print(X_test.shape[0], 'test samples')
# convert class vectors to binary class matrices
Y_train = np_utils.to_categorical(y_train, NB_CLASSES)
Y_test = np_utils.to_categorical(y_test, NB_CLASSES)
# M_HIDDEN hidden layers
# 10 outputs
# final stage is softmax
model = Sequential()
model.add(Dense(N_HIDDEN, input_shape=(RESHAPED,)))
model.add(Activation('relu'))
model.add(Dense(N_HIDDEN))
model.add(Activation('relu'))
model.add(Dense(NB_CLASSES))
model.add(Activation('softmax'))
model.summary()
model.compile(loss='categorical_crossentropy',
optimizer=OPTIMIZER,
metrics=['accuracy'])
history = model.fit(X_train, Y_train,
batch_size=BATCH_SIZE, epochs=NB_EPOCH,
verbose=VERBOSE, validation_split=VALIDATION_SPLIT)
score = model.evaluate(X_test, Y_test, verbose=VERBOSE)
print("Test score:", score[0])
print('Test accuracy:', score[1])

60000 train samples
10000 test samples
Model: "sequential_3"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_7 (Dense)              (None, 128)               100480    
_________________________________________________________________
activation_7 (Activation)    (None, 128)               0         
_________________________________________________________________
dense_8 (Dense)              (None, 128)               16512     
_________________________________________________________________
activation_8 (Activation)    (None, 128)               0         
_________________________________________________________________
dense_9 (Dense)              (None, 10)                1290      
_________________________________________________________________
activation_9 (Activation)    (None, 10)                0         
Total params: 118,282
Trainable params: 118,282
Non-trainable params: 0
_________

48000/48000 [==============================] - 4s 87us/step - loss: 0.0487 - accuracy: 0.9873 - val_loss: 0.0959 - val_accuracy: 0.9728
Epoch 102/150
48000/48000 [==============================] - 5s 112us/step - loss: 0.0481 - accuracy: 0.9876 - val_loss: 0.0946 - val_accuracy: 0.9727
Epoch 103/150
48000/48000 [==============================] - 6s 118us/step - loss: 0.0474 - accuracy: 0.9880 - val_loss: 0.0944 - val_accuracy: 0.9732
Epoch 104/150
48000/48000 [==============================] - 5s 108us/step - loss: 0.0467 - accuracy: 0.9884 - val_loss: 0.0942 - val_accuracy: 0.9737
Epoch 105/150
48000/48000 [==============================] - 5s 107us/step - loss: 0.0462 - accuracy: 0.9883 - val_loss: 0.0948 - val_accuracy: 0.9743
Epoch 106/150
48000/48000 [==============================] - 6s 118us/step - loss: 0.0457 - accuracy: 0.9883 - val_loss: 0.0940 - val_accuracy: 0.9740
Epoch 107/150
48000/48000 [==============================] - 5s 108us/step - loss: 0.0451 - accuracy: 0.9885 

The first code block is the copied code from the text.<br>
The results from running the first code block with 2 hidden layers and 20 epochs are as follows:<br>
We get 94.56% accuracy on training, 94.98% on validation and 94.63% on the test.<br>
<br>
According to out text, it is significant to note that training accuracy should be higher than the test accuracy, otherwise we are not training long enough (Gulli, A. & Pal, S., 2017)<br>
<br>
Therefore, I have provided two experiments where the parameter I chose to modify is the number of epochs (NB_EPOCH) to<br> determine whether or not we can increase the accuracy rate.<br>
The following are the results of the two experiments.<br>
<br>
Experiment 1 (code block 2): Increased epochs to 50 with 2 hidden layers<br>
We get 97.28% accuracy on training, 96.61% on validation, and 96.57% on the test.<br>
<br>
Also, we see from experiment 1 that the training accuracy is now higher than the test accuracy, which tells us we are starting to train long enough.<br>
<br>
Experiment 2 (code block 3): Increased epochs to 150 with 2 hidden layers<br>
We get 99.45% accuracy on training, 97.58% on validation, and 97.48% on the test.<br>
<br>
By conducting these experiments, we see that by increasing the number of epochs each time, we have improved the training,<br> validation, and test accuracy, and that the model has become well trained.<br>
<br>
Experiment 1 resulted in a 2.72% increase in training accuracy, 1.63% increase on validation and 1.94% increase on the test<br> from the initial code block.<br>
<br>
Additionally, experiment 2 resulted in an additional 2.17% increase in training accuracy, .97% increase on validation and .91% increase on the test.<br>
<br>
The reason why we are seeing an increase in the training, validation, and tests as the number of epochs are increased is<br> because we are allowing the model to be exposed to the data an increased number of times, which improves the accuracy as it<br> learns from the data. Every new epoch is a new opportunity for the model to see the data and learn from it, and therefore is<br> able to adjust its weights and biases, minimizing loss.<br>
<br>
After having conducted experiment 2 and see such a high percentage on training and test, we can conclude that there is no<br> longer a need to increase the number of epochs as this would result in a higher amount of computational time with little to no<br> additional learning occuring.<br>
<br>
References:<br>
<br>
Gulli, A., & Pal, S. (2017). Deep learning with keras: Get to grips with the basics of keras to implement fast and efficient deep-learning models. Packt Publishing, Limited.
